domain/application/service/application.go

```diff
@@ -533,54 +671,172 @@ func (s *ApplicationService) RegisterCAASUnit(ctx context.Context, appName strin
+		if err := s.st.UpdateUnitContainer(ctx, args.UnitName, cloudContainer); err != nil {
+			return errors.Annotatef(err, "updating unit %q", args.UnitName)
+		}
+
+		// We want to transition to using unit UUID instead of name.
+		unitUUID, err := s.st.GetUnitUUID(ctx, args.UnitName)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		return s.st.SetUnitPassword(ctx, unitUUID, application.PasswordInfo{
+			PasswordHash:  args.PasswordHash,
+			HashAlgorithm: application.HashAlgorithmSHA256,
+		})
 	})
 	return errors.Annotatef(err, "saving caas unit %q", args.UnitName)
 }
 
 func (s *ApplicationService) insertCAASUnit(
-	ctx domain.AtomicContext, appID coreapplication.ID, orderedID int, args application.UpsertUnitArg,
+	ctx domain.AtomicContext, appID coreapplication.ID, orderedID int, arg application.InsertUnitArg,
 ) error {
 	appScale, err := s.st.GetApplicationScaleState(ctx, appID)
 	if err != nil {
 		return errors.Annotatef(err, "getting application scale state for app %q", appID)
 	}
 	if orderedID >= appScale.Scale ||
 		(appScale.Scaling && orderedID >= appScale.ScaleTarget) {
-		return fmt.Errorf("unrequired unit %s is not assigned%w", args.UnitName, errors.Hide(applicationerrors.UnitNotAssigned))
+		return fmt.Errorf("unrequired unit %s is not assigned%w", arg.UnitName, errors.Hide(applicationerrors.UnitNotAssigned))
+	}
+	return s.st.InsertUnit(ctx, appID, arg)
+}
+
+// UpdateCAASUnit updates the specified CAAS unit, returning as error
+// satisfying applicationerrors.ApplicationNotAlive if the unit's
+// application is not alive.
+func (s *ApplicationService) UpdateCAASUnit(ctx context.Context, unitName string, params UpdateCAASUnitParams) error {
+	var cloudContainer *application.CloudContainer
+	if params.ProviderId != nil {
+		cloudContainerParams := CloudContainerParams{
+			ProviderId: *params.ProviderId,
+			Ports:      params.Ports,
+		}
+		if params.Address != nil {
+			addr := network.NewSpaceAddress(*params.Address, network.WithScope(network.ScopeMachineLocal))
+			cloudContainerParams.Address = &addr
+			origin := network.OriginProvider
+			cloudContainerParams.AddressOrigin = &origin
+		}
+		cloudContainer = makeCloudContainerArg(unitName, cloudContainerParams)
 	}
-	return s.st.InsertUnit(ctx, appID, args)
+	err := s.st.RunAtomic(ctx, func(ctx domain.AtomicContext) error {
+		appName, _ := names.UnitApplication(unitName)
+		_, appLife, err := s.st.GetApplicationLife(ctx, appName)
+		if err != nil {
+			return fmt.Errorf("getting application %q life: %w", appName, err)
+		}
+		if appLife != life.Alive {
+			return fmt.Errorf("application %q is not alive%w", appName, errors.Hide(applicationerrors.ApplicationNotAlive))
+		}
+
+		if cloudContainer != nil {
+			if err := s.st.UpdateUnitContainer(ctx, unitName, cloudContainer); err != nil {
+				return errors.Annotatef(err, "updating cloud container %q", unitName)
+			}
+		}
+		// We want to transition to using unit UUID instead of name.
```

Comment:
As stated in the PR description, we need to use UUIDs not natural keys but it's a big uplift and it all can't be done in a feature PR. It will be done in a followup in a very short timeframe.